File: test-suite/handwritten-src/js/AsyncTest.js
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

async function asyncFunc() {
    await sleep(10);
    return 36;
}

async function asyncException() {
    await sleep(10);
    throw new Error("123");
}

class AsyncInterfaceImpl {
    async futureRoundtrip(f) {
        const i = await f;
        return i.toString();
    }
}

class AsyncTest {
    constructor(module) {
        this.module = module;
    }
    async testConsumeNativeFuture() {
        const r = await this.module.testsuite.TestHelpers.getAsyncResult();
        assertEq(r, 42);
    }

    async testConsumeNativeFutureAltSyntax() {
        return this.module.testsuite.TestHelpers.getAsyncResult().then(res => assertEq(res, 42)).catch(e => assertTrue(false));
    }

    async testFutureRoundtrip() {
        const s = await this.module.testsuite.TestHelpers.futureRoundtrip(asyncFunc());
        const r = parseInt(s);
        assertEq(r, 36);
    }

    async testVoidRoundTrip() {
        await this.m.testsuite.TestHelpers.voidAsyncMethod(sleep(10));
    }

    async testOptionalFutureUnsetValue() {
        assertUndefined(await this.m.testsuite.TestHelpers.addOneIfPresent(Promise.resolve(undefined)));
    }

    async testOptionalFutureSetValue() {
        assertEq(await this.m.testsuite.TestHelpers.addOneIfPresent(Promise.resolve(10)), 11);
    }

    async testFutureRoundtripWithException() {
        var s = null;
        try {
            const s = await this.module.testsuite.TestHelpers.futureRoundtrip(asyncException());
            const r = parseInt(s);
            assertEq(r, 36);
        } catch (e) {
            s = e.message;
        }
        console.log(s);
        assertEq(s, "123");
    }

    async testFutureRoundtripBackwards() {
        const s = await this.module.testsuite.TestHelpers.checkAsyncInterface(new AsyncInterfaceImpl());
        assertEq(s, "36");
    }

    async testFutureComposition() {
        const s = await this.module.testsuite.TestHelpers.checkAsyncComposition(new AsyncInterfaceImpl());
        assertEq(s, "42");
    }

    async testEarlyThrow() {
        const r = this.module.testsuite.TestHelpers.asyncEarlyThrow();
        let s = null;
        try {
            await r;
        } catch (e) {
            s = e.message;
        }
        assertEq(s, "C++: error");
    }

    async testReturnExceptionString_valid() {
        const result = await this.module.testsuite.TestHelpers.returnExceptionString(Promise.reject(new Error("hello")));
        assertTrue(result.startsWith("Error: hello"));
    }

    async testReturnExceptionString_null() {
        const result = await this.module.testsuite.TestHelpers.returnExceptionString(Promise.reject(null));
        assertEq(result, "JS promise rejected with non-error type");
    }

    async testReturnExceptionString_undefined() {
        const result = await this.module.testsuite.TestHelpers.returnExceptionString(Promise.reject(undefined));
        assertEq(result, "JS promise rejected with non-error type");
    }
}

allTests.push(AsyncTest);

File: test-suite/handwritten-src/swift/AsyncTest.swift
import XCTest
import DjinniSupport
@testable import TestSuite

final class AsyncTest: XCTestCase {

    class AsyncInterfaceImpl: AsyncInterface {
        func futureRoundtrip(_ f: DJFuture<Int32>) throws -> DJFuture<String> {
            var token: Cancellable? = nil
            return DJFuture() { promise in
                token = f.getResult { result in
                    switch result {
                    case .success(let value):
                        promise(.success(String(value)))
                    case .failure:
                        XCTFail()
                        token = withExtendedLifetime(token) {nil}
                    }
                }
            }
        }
    }
    
    func testConsumeNativeFuture() async throws {
        let f = try TestHelpers_statics.getAsyncResult()
        let i = try await f.value
        XCTAssertEqual(i, 42)
    }

    func testFutureRoundtrip() async throws {
        var p: DJFuture<String>.Promise? = nil
        let f = DJFuture<String>() { promise in
            p = promise
        }
        var cancellable: Cancellable? = nil
        let f2 = DJFuture<Int32>() { promise in
            cancellable = f.getResult { result in
                switch result {
                case .failure:
                    XCTFail()
                case .success(let value):
                    promise(.success(Int32(value)!))
                }
            }
        }
        let f3 = try TestHelpers_statics.futureRoundtrip(f2)
        p!(.success("36"))
        let res = try await f3.value
        XCTAssertEqual(res, "36")
        cancellable = withExtendedLifetime(cancellable) {nil}
    }

    func testVoidRoundtrip() async throws {
        let f = DJFuture<Void>() { promise in
            promise(.success(()))
        }
        let f1 = try TestHelpers_statics.voidAsyncMethod(f)
        try await f1.value
    }

    func testFutureRoundtripWithException() async throws {
        var p: DJFuture<String>.Promise? = nil
        let f = DJFuture<String>() { promise in
            p = promise
        }
        var cancellable: Cancellable? = nil
        let f2 = DJFuture<Int32>() { promise in
            cancellable = f.getResult { result in
                switch result {
                case .failure:
                    XCTFail()
                case .success(let value):
                    promise(.failure(DjinniError("123")))
                }
            }
        }
        let f3 = try TestHelpers_statics.futureRoundtrip(f2)
        p!(.success("36"))
        var s: String? = nil
        do {
            let _ = try await f3.value
        } catch let e as DjinniError {
            s = e.errorMessage
        }
        XCTAssertEqual(s, "123")
        cancellable = withExtendedLifetime(cancellable){nil}
    }

    func testFutureRoundtripBackwards() async throws {
        let s = try await TestHelpers_statics.checkAsyncInterface(AsyncInterfaceImpl()).value
        XCTAssertEqual(s, "36")
    }

    func testFutureComposition() async throws {
        let s = try await TestHelpers_statics.checkAsyncComposition(AsyncInterfaceImpl()).value
        XCTAssertEqual(s, "42")
    }

    func testOptionalFutureUnsetValue() async throws {
        let input = DJFuture<Optional<Int32>> { promise in
            promise(.success(nil))
        }
        let output = try await TestHelpers_statics.addOneIfPresent(input).value
        XCTAssertNil(output)
    }

    func testOptionalFutureSetValue() async throws {
        let input = DJFuture<Optional<Int32>> { promise in
            promise(.success(10))
        }
        let output = try await TestHelpers_statics.addOneIfPresent(input).value
        XCTAssertEqual(output, 11)
    }
}
